fluent-langneg/src/locale.ts

```diff
@@ -39,29 +41,55 @@ export class Locale {
    * properly parsed as `en-*-US-*`.
    */
   constructor(locale: string) {
-    const result = localeRe.exec(locale.replace(/_/g, "-"));
-    if (!result) {
+    let result;
+    let normalized = convertMasks(locale.replace(/_/g, "-"));
+    try {
+      result = new Intl.Locale(normalized);
+    } catch (e) {
       this.isWellFormed = false;
+      this.language = "und";
       return;
     }
 
-    let [, language, script, region, variant] = result;
+    this.language = result.language || "und";
+    this.script = result.script;
+    this.region = result.region;
 
-    if (language) {
-      this.language = language.toLowerCase();
+    let visiblelangTagLength = getVisibleLangTagLength(this.language, this.script, this.region);
+
+    if (normalized.length > visiblelangTagLength) {
+      let extStart = getExtensionStart(locale);
+      this.variant = locale.substring(visiblelangTagLength + 1, extStart);
+    }
+
+    this.isWellFormed = true;
+  }
+
+  static fromComponents({language, script, region, variant}: {
```

Comment:
Could/should this also use `Intl.Locale`? The original included some normalizations that are included in `Intl.Locale` processing, but they're dropped in these changes. Code here could construct an options bag to pass as the second arg of the Intl constructor, and then read the language/script/region from the result.